packages/daemon/src/service/trace-manager.ts

```diff
@@ -0,0 +1,208 @@
+import { Transform, TransformCallback } from 'stream';
+import { EventEmitter } from 'events';
+import { open, close, write } from 'fs-extra';
+import { provide, scope, ScopeEnum } from 'midway';
+import { StringDecoder } from 'string_decoder';
+import { generateId, PipelineStatus, PluginTypeI } from '@pipcook/pipcook-core';
+
+export type PipcookEventType = 'log' | 'job_status';
+
+export interface JobStatusChangeEvent {
+  jobStatus: PipelineStatus;
+  step?: PluginTypeI;
+  stepAction?: 'start' | 'end';
+}
+
+export interface LogEvent {
+  level: string;
+  data: string;
+}
+
+export type PipecookEvent = JobStatusChangeEvent | LogEvent;
+/**
+ * tracer for plugin installing and pipeline running.
+ */
+export class Tracer {
+  // log id
+  id: string;
+  // stdout stream for log pipe
+  private stdout: LogPassthrough;
+  // stderr stream for log pipe
+  private stderr: LogPassthrough;
+  // event emitter
+  private emitter: EventEmitter;
+
+  private opts?: LogOptions;
+  private fdOut: number;
+  private fdErr: number;
+  constructor(opts?: LogOptions) {
+    this.id = generateId();
+    this.emitter = new EventEmitter();
+    this.opts = opts;
+  }
+
+  /**
+   * init loggers
+   */
+  async initLogger() {
```

Comment:
Could you please explain a bit the relationship between logger with trace?

Reply:
Updated, put the annotation on the class.

Reply:
I have commented some, and the logger should be a module(class) based on the Trace.

Reply:
Sorry, I doesn't get it. What you mean `the logger should be a module(class) based on the Trace`? In the current implementation, the loggers are part of the tracer.

Reply:
> the loggers are part of the tracer.

This looks weird, why does the tracer own the log-related props?